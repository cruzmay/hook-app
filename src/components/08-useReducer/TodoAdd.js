import React from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoAdd = ({ handleAddToDo }) => {
    
    const [{description}, handleInputChange, reset] = useForm({description: ''})

    const handleSubmit = (e) => {
        e.preventDefault()

        if(description.trim().length <= 1){
            return
        }

        const newTodo = {
            id : new Date().getTime(),
            desc: description,
            done: false
        }

        handleAddToDo(newTodo)
        reset()

    }

    return (
        <>
            <h3>Agregar</h3>
            <form 
                onSubmit={handleSubmit}
                className="form-group col-12 m-4 d-flex justify-content-center"
            >
                <input
                    className="col-9"
                    type="test"
                    name="description"
                    placeholder="aprender"
                    autoComplete="off"
                    value={description}
                    onChange={handleInputChange}
                />
                <button
                    type="submit"
                    className="btn btn-primary col-3 ms-4"
                >
                    Agregar
                </button>
            </form>
        </>
    )
}
