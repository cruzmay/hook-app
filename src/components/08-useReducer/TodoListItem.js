import React from 'react'

export const TodoListItem = ({id, desc, done, handleDelete, handleToogle, index}) => {
    return (
        <li 
        
        className="list-group-item col-12 d-flex justify-content-between align-items-center"
        >
            <p 
                className={`col-4 m-0 task ${ done && 'tacha'} `}
                onClick={()=>handleToogle(id)}
            >
            {index + 1}.- {desc}
            </p>
            <button 
                className="btn btn-small btn-danger"
                onClick={() => handleDelete(id)}
            >
                Borrar
            </button>
        </li>
    )
}
