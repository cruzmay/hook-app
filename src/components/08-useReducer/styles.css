body {
  padding: 50px;
}
.task {
  cursor: pointer;
}
.tacha {
  text-decoration: line-through;
}
