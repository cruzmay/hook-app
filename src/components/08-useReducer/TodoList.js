import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({todos, handleDelete, handleToogle}) => {
    return (
        <ul>
            {
                todos.map((todo, i)=> 
                    <TodoListItem 
                        {...todo}
                        key={todo.id}
                        handleDelete={handleDelete}
                        handleToogle={handleToogle}
                        index = {i}
                    />
                )
            }
            
        </ul>
    )
}
