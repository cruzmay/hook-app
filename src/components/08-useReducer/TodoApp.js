import React, { useEffect, useReducer } from 'react';
import './styles.css'
import { todoReducer } from './todoReducer';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

 const init = () => {
     return JSON.parse(localStorage.getItem('todos'))
 }

export const TodoApp = () => {
    
    const [todos, dispatch] = useReducer(todoReducer, [], init)

    useEffect(()=>{
        localStorage.setItem('todos', JSON.stringify(todos))
    },[todos])

    const handleDelete = (todoId) => {
        console.log(todoId)
        const deleteAction = {
            type: 'delete',
            payload: todoId
        }
        dispatch(deleteAction)
    }

    const handleToogle = (todoId) => {
        // console.log(todoId)
        dispatch({
            type: 'toogle',
            payload: todoId
        })
    }

    const handleAddToDo = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        })
    }

    return (
        <div>
            <h1>TodoApp ({todos.length})</h1>
            <hr/>
            <TodoAdd
                handleAddToDo={handleAddToDo}
            />
            <TodoList
                todos = {todos}
                handleDelete={handleDelete}
                handleToogle={handleToogle}
            />
        </div>
    )
}
