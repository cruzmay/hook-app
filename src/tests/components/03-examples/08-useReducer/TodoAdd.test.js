import { shallow } from 'enzyme';
import React from 'react';
import { TodoAdd } from "../../../../components/08-useReducer/TodoAdd"

describe('pruebas en <TodoAdd/>', () => {

    const handleAddToDo = jest.fn()

    const wrapper = shallow(
            <TodoAdd 
            handleAddToDo = { handleAddToDo }
            />
        )

    test('debe de mostrarse correctamente', () => {
         expect(wrapper).toMatchSnapshot()
    })
    test('No debe de llamar handleAddToDo', () => {
        const formSubmit = wrapper.find('form').prop('onSubmit')
        formSubmit({preventDefault(){}})
        expect(handleAddToDo).toHaveBeenCalledTimes(0)
    })
    test('debe de llamar la funcion handleAddToDo', () => {
        //con un argumento/called
        const value = 'Hacer la tarea'
        const input = wrapper.find('input')
        input.simulate('change', { target:{ value, name : 'description'  } })
        console.log(input.props())
        wrapper.find('button').simulate('click')
        const form = wrapper.find('form')
        form.simulate('submit', {preventDefault: ()=>{}} )

        expect(handleAddToDo).toHaveBeenCalledTimes(1)
        expect(handleAddToDo).toHaveBeenCalledWith(expect.any(Object))
        expect(handleAddToDo).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc: value,
            done: false 
        })
        expect(wrapper.find('input').prop('value')).toBe('')
    })
    
    
})
