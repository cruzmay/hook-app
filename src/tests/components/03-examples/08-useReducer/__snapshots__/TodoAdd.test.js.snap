// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <TodoAdd/> debe de mostrarse correctamente 1`] = `
<Fragment>
  <h3>
    Agregar
  </h3>
  <form
    className="form-group col-12 m-4 d-flex justify-content-center"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="col-9"
      name="description"
      onChange={[Function]}
      placeholder="aprender"
      type="test"
      value=""
    />
    <button
      className="btn btn-primary col-3 ms-4"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
